File: packages/cli/src/commands/frozen.ts
import { Command } from "commander";
import * as path from "node:path";
import {
  exportFrozenBundle,
  importFrozenBundle,
  verifyFrozenBundle,
} from "../core/frozen.js";
import { CliError } from "../core/exit.js";

interface FrozenExportOptions {
  frozen?: boolean;
  root?: string;
  out?: string;
  json?: boolean;
}

interface FrozenImportOptions {
  frozen?: boolean;
  target?: string;
  dryRun?: boolean;
  merge?: boolean;
  force?: boolean;
  json?: boolean;
}

interface FrozenVerifyOptions {
  json?: boolean;
}

function printJson(data: unknown): void {
  console.log(JSON.stringify(data, null, 2));
}

export function frozenExportCommand(): Command {
  return new Command("export")
    .description("Export a frozen x-harness bundle")
    .option("--frozen", "Export frozen harness bundle", false)
    .requiredOption("--out <path>", "Output bundle path")
    .option("--root <path>", "Repository root", process.cwd())
    .option("--json", "Output JSON instead of text", false)
    .action(async (opts: FrozenExportOptions) => {
      if (!opts.frozen) {
        throw new CliError("export currently requires --frozen", 2);
      }
      const result = await exportFrozenBundle({
        root: path.resolve(opts.root ?? process.cwd()),
        out: opts.out as string,
      });
      if (opts.json) printJson(result);
      else {
        console.log(`frozen bundle written: ${result.out}`);
        console.log(`files: ${result.file_count}`);
      }
    });
}

export function frozenImportCommand(): Command {
  return new Command("import")
    .description("Import a frozen x-harness bundle")
    .argument("<bundle>", "Frozen bundle path")
    .option("--frozen", "Import frozen harness bundle", false)
    .requiredOption("--target <path>", "Target repository path")
    .option("--dry-run", "Preview import without writing", true)
    .option("--merge", "Write missing files but preserve existing files", false)
    .option("--force", "Overwrite existing files", false)
    .option("--json", "Output JSON instead of text", false)
    .action(async (bundle: string, opts: FrozenImportOptions) => {
      if (!opts.frozen) {
        throw new CliError("import currently requires --frozen", 2);
      }
      const dryRun = !opts.merge && !opts.force;
      const result = await importFrozenBundle({
        bundlePath: bundle,
        target: opts.target as string,
        dryRun,
        merge: opts.merge,
        force: opts.force,
      });
      if (opts.json) printJson(result);
      else {
        console.log(
          dryRun
            ? `frozen import dry-run: ${result.planned.length} file(s)`
            : `frozen import wrote ${result.written.length} file(s)`
        );
        for (const conflict of result.conflicts) {
          console.log(`conflict: ${conflict}`);
        }
      }
      if (!result.ok) {
        throw new CliError("frozen import failed", 1);
      }
    });
}

export function frozenCommand(): Command {
  const frozen = new Command("frozen").description(
    "Verify frozen x-harness transfer bundles"
  );
  frozen
    .command("verify")
    .description("Verify frozen bundle manifest and checksums")
    .argument("<bundle>", "Frozen bundle path")
    .option("--json", "Output JSON instead of text", false)
    .action(async (bundle: string, opts: FrozenVerifyOptions) => {
      const result = await verifyFrozenBundle(bundle);
      if (opts.json) printJson(result);
      else if (result.ok) {
        console.log(`frozen bundle valid: ${result.bundle_path}`);
        console.log(`files: ${result.file_count}`);
      } else {
        console.log("frozen bundle invalid:");
        for (const error of result.errors) console.log(`- ${error}`);
      }
      if (!result.ok) {
        throw new CliError("frozen verify failed", 1);
      }
    });
  return frozen;
}

File: internal/cli/frozen.go
package cli

import (
	"encoding/json"
	"fmt"
	"io"
	"path/filepath"
	"strings"

	"github.com/BrianNguyen29/x-harness/internal/frozen"
)

func handleFrozen(args []string, stdout, stderr io.Writer) int {
	if len(args) == 0 {
		fmt.Fprintln(stderr, "frozen requires a subcommand: export, import, verify")
		return ExitUsage
	}

	switch args[0] {
	case "export":
		return handleFrozenExport(args[1:], stdout, stderr)
	case "import":
		return handleFrozenImport(args[1:], stdout, stderr)
	case "verify":
		return handleFrozenVerify(args[1:], stdout, stderr)
	default:
		fmt.Fprintf(stderr, "unknown frozen subcommand: %s\n", args[0])
		return ExitUsage
	}
}

func handleFrozenExport(args []string, stdout, stderr io.Writer) int {
	root := "."
	out := ""
	frozenFlag := false
	jsonMode := false

	for i := 0; i < len(args); i++ {
		switch args[i] {
		case "--frozen":
			frozenFlag = true
		case "--out":
			if i+1 >= len(args) {
				fmt.Fprintln(stderr, "error: --out requires a value")
				return ExitUsage
			}
			out = args[i+1]
			i++
		case "--root":
			if i+1 >= len(args) {
				fmt.Fprintln(stderr, "error: --root requires a value")
				return ExitUsage
			}
			root = args[i+1]
			i++
		case "--json":
			jsonMode = true
		default:
			if strings.HasPrefix(args[i], "-") {
				fmt.Fprintf(stderr, "unknown flag: %s\n", args[i])
				return ExitUsage
			}
			fmt.Fprintf(stderr, "unexpected argument: %s\n", args[i])
			return ExitUsage
		}
	}

	if !frozenFlag {
		fmt.Fprintln(stderr, "export currently requires --frozen")
		return ExitUsage
	}
	if out == "" {
		fmt.Fprintln(stderr, "error: --out <path> is required")
		return ExitUsage
	}

	root, _ = filepath.Abs(root)
	result, err := frozen.ExportFrozenBundle(root, out)
	if err != nil {
		fmt.Fprintf(stderr, "error: %v\n", err)
		return ExitError
	}

	if jsonMode {
		data, _ := json.MarshalIndent(result, "", "  ")
		fmt.Fprintln(stdout, string(data))
	} else {
		fmt.Fprintf(stdout, "frozen bundle written: %s\n", result.Out)
		fmt.Fprintf(stdout, "files: %d\n", result.FileCount)
	}
	return ExitOK
}

func handleFrozenImport(args []string, stdout, stderr io.Writer) int {
	bundle := ""
	target := ""
	frozenFlag := false
	dryRun := true
	merge := false
	force := false
	jsonMode := false

	for i := 0; i < len(args); i++ {
		switch args[i] {
		case "--frozen":
			frozenFlag = true
		case "--target":
			if i+1 >= len(args) {
				fmt.Fprintln(stderr, "error: --target requires a value")
				return ExitUsage
			}
			target = args[i+1]
			i++
		case "--dry-run":
			dryRun = true
		case "--no-dry-run":
			dryRun = false
		case "--merge":
			merge = true
		case "--force":
			force = true
		case "--json":
			jsonMode = true
		default:
			if strings.HasPrefix(args[i], "-") {
				fmt.Fprintf(stderr, "unknown flag: %s\n", args[i])
				return ExitUsage
			}
			if bundle == "" {
				bundle = args[i]
			} else {
				fmt.Fprintf(stderr, "unexpected argument: %s\n", args[i])
				return ExitUsage
			}
		}
	}

	if !frozenFlag {
		fmt.Fprintln(stderr, "import currently requires --frozen")
		return ExitUsage
	}
	if bundle == "" {
		fmt.Fprintln(stderr, "error: bundle path is required")
		return ExitUsage
	}
	if target == "" {
		fmt.Fprintln(stderr, "error: --target <path> is required")
		return ExitUsage
	}

	if merge || force {
		dryRun = false
	}

	result, err := frozen.ImportFrozenBundle(bundle, target, dryRun, merge, force)
	if err != nil {
		fmt.Fprintf(stderr, "error: %v\n", err)
		return ExitError
	}

	if jsonMode {
		data, _ := json.MarshalIndent(result, "", "  ")
		fmt.Fprintln(stdout, string(data))
	} else {
		if dryRun {
			fmt.Fprintf(stdout, "frozen import dry-run: %d file(s)\n", len(result.Planned))
		} else {
			fmt.Fprintf(stdout, "frozen import wrote %d file(s)\n", len(result.Written))
		}
		for _, conflict := range result.Conflicts {
			fmt.Fprintf(stdout, "conflict: %s\n", conflict)
		}
	}

	if !result.OK {
		fmt.Fprintln(stderr, "frozen import failed")
		return ExitError
	}
	return ExitOK
}

func handleFrozenVerify(args []string, stdout, stderr io.Writer) int {
	bundle := ""
	jsonMode := false

	for i := 0; i < len(args); i++ {
		switch args[i] {
		case "--json":
			jsonMode = true
		default:
			if strings.HasPrefix(args[i], "-") {
				fmt.Fprintf(stderr, "unknown flag: %s\n", args[i])
				return ExitUsage
			}
			if bundle == "" {
				bundle = args[i]
			} else {
				fmt.Fprintf(stderr, "unexpected argument: %s\n", args[i])
				return ExitUsage
			}
		}
	}

	if bundle == "" {
		fmt.Fprintln(stderr, "error: bundle path is required")
		return ExitUsage
	}

	result, err := frozen.VerifyFrozenBundle(bundle)
	if err != nil {
		fmt.Fprintf(stderr, "error: %v\n", err)
		return ExitError
	}

	if jsonMode {
		data, _ := json.MarshalIndent(result, "", "  ")
		fmt.Fprintln(stdout, string(data))
	} else if result.OK {
		fmt.Fprintf(stdout, "frozen bundle valid: %s\n", result.BundlePath)
		fmt.Fprintf(stdout, "files: %d\n", result.FileCount)
	} else {
		fmt.Fprintln(stdout, "frozen bundle invalid:")
		for _, e := range result.Errors {
			fmt.Fprintf(stdout, "- %s\n", e)
		}
	}

	if !result.OK {
		fmt.Fprintln(stderr, "frozen verify failed")
		return ExitError
	}
	return ExitOK
}
